### PR TITLE
feat: Add search functionality while selecting icons

### LINKED ISSUE #43: [FEATURE] Search filter for icons
### Description

**Problem**
When selecting icons for adding to the link or milestones we have to scroll through this entire list which is quite cumbersome. 

**Potential solution**
It would be great if we can take input from user and create a search filter instead. This will reduce time to search for the desired icon. They can just type in the name of the icon and matched icons will show up to select from.

### Screenshots

Screenshot 2023-02-24 184618

### PR DESCRIPTION
**Pull request by PratikDey**
## 🛠️ Fixes Issue

Closes #43 

## 👨‍💻 Changes proposed
Added autocomplete feature for selecting icons

## ✔️ Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## 📷 Screenshots

### CODE REVIEW

**Review by Ananya2001-an — DISMISSED** [review]
Looks awesome! Thanks for your contribution 👍

**Comment by Ananya2001-an** [comment]
Hey @Pradumnasaraf can you also review the changes.

**Comment by Ananya2001-an** [comment]
There is just one issue that I think we need to look into that is when we show the icons users might not know that they can actually search by typing. It shows a cursor but it just might not catch their eye. What do you think @PratikDey @Pradumnasaraf ?

image

https://www.awesomescreenshot.com/video/15244139?key=c020dc53af47746365f77df189e7f3c3

**Comment by Pradumnasaraf** [comment]
Hey, 
First, I didn't know we had a type of type: "autocomplete"`. Learned something new. 

But I think, we can't use this one because the prompt message get disappeared and the user will not understand what should they do also the whole terminal is filled up with options.

So we need need to implement this, as mentioned in the issue.
https://www.npmjs.com/package/enquirer#autocomplete-prompt
Screenshot 2023-03-02 at 11 56 06 AM

**Review by Pradumnasaraf — COMMENTED** [review]
Hey, 
First, I didn't know we had a type of `type: "autocomplete"`. Learned something new. 

But I think, we can't use this one because the prompt message get disappeared and the user will not understand what should they do also the whole terminal is filled up with options.

So we need need to implement this, as mentioned in the issue.
https://www.npmjs.com/package/enquirer#autocomplete-prompt
Screenshot 2023-03-02 at 11 56 06 AM

**Comment by Ananya2001-an** [comment]
> Hey, First, I didn't know we had a type of `type: "autocomplete"`. Learned something new.
> 
> But I think, we can't use this one because the prompt message get disappeared and the user will not understand what should they do also the whole terminal is filled up with options.
> 
> So we need need to implement this, as mentioned in the issue. https://www.npmjs.com/package/enquirer#autocomplete-prompt Screenshot 2023-03-02 at 11 56 06 AM

Yeah I think that will be better as users will not know otherwise that they can type and search.

**Comment by PratikDey** [comment]
Hi, @Ananya2001-an and @Pradumnasaraf , I have actually tried implementing Autocomplete as it is in documentation but I am getting errors trying to implement like that. If you folks can give me some pointers on how should I go on implementing this then that would be much helpful. Also sorry, I don't have much knowledge on javascript.

**Comment by Ananya2001-an** [comment]
> Hi, @Ananya2001-an and @Pradumnasaraf , I have actually tried implementing Autocomplete as it is in documentation but I am getting errors trying to implement like that. If you folks can give me some pointers on how should I go on implementing this then that would be much helpful. Also sorry, I don't have much knowledge on javascript.

No worries @PratikDey Can you tell us the exact problem that you are facing while implementing it. Maybe we can guide you then. I think after creating an instance of Autocomplete you can try writing the prompt.run() command after all other prompts are asked. Just try and let me know if you are facing any issues. 

Also don't worry about JavaScript. Everyone is learning something new here. I sucked at JS too at one point. It's all about trial and error. ☺

**Comment by PratikDey** [comment]
Hi, when I am trying to run autocomplete by calling it as a function I am getting this error
image
and if I am calling the prompt in autocomplete function I am getting error like this
image

This is another approach I tried
image

Am I making any mistakes here?

**Comment by Ananya2001-an** [comment]
Following the last screenshot move the autocomplete variable outside the while loop and also remove the prompt where you have given choices as autocomplete. Instead call autocomplete.run() just after all the prompts. I hope you are understanding what I am saying ?

**Comment by PratikDey** [comment]
Yes, it is working. Thank you. 
Sorry, I am not so good in Javascript, that is why I had to disturb you folks so much.
I'll be submitting a PR soon

**Comment by Ananya2001-an** [comment]
> Yes, it is working. Thank you. Sorry, I am not so good in Javascript, that is why I had to disturb you folks so much. I'll be submitting a PR soon

No issues. We are maintainers for a reason 😉

**Comment by Ananya2001-an** [comment]
Also @PratikDey no need to raise another PR you can make new commits to the same one.

**Comment by Pradumnasaraf** [comment]
Yes, you can push the changes to this PR as you move forward (even if it's broken) so that we can see and give you feedback and test if you get stuck

**Comment by PratikDey** [comment]
hi, this PR is ready for review

**Comment by Ananya2001-an** [comment]
Hey @PratikDey I tested the code and there are some issues:
1. First of all the autocomplete variable should be created inside the addLinks/updateLinks function so that choices list is not empty else it will give undefined error.
2. Also the icon is not getting saved inside the link object for the JSON profile in this way. Have you tested the code properly?

I will suggest you to play around with the code a bit and see what impact it has on the results so that you can get a better understanding. Also I feel you can do something like this that I tried and just apply the same logic everywhere else. Just make sure to import AutoComplete module from enquirer package.

```
async function addlinks(bool) {
  icons = await geticons();
  const autocomplete = new AutoComplete({
    name: 'icon',
    message: 'Choose an icon (Search to see more options)',
    limit: 10,
    choices: icons
  })
  while (bool) {
    let answers = await prompt([
      {
        type: "input",
        name: "name",
        message: "What is the name of the link?",
      },
      {
        type: "input",
        name: "url",
        message: "What is the URL of the link?",
      }
    ]);
    await autocomplete.run()
    const confirm = await prompt({
      type: "confirm",
      name: "addLink",
      message: "Do you want to add another link?",
    })
    links.push({
      name: answers.name,
      url: answers.url,
      icon: autocomplete.state.input,
    });
    if (!confirm.addLink) {
      return links;
    }
  }
}
```

**Comment by Pradumnasaraf** [comment]
Weirdly, I am losing notifications on GItHub. Thanks for answering, @Ananya2001-an.

Yes I agree, we need to break code the icon part into 3 section

```js
await prompt ([ ])
await autocomplete.run()
await prompt ([ ])

// the push the inputs

```

**Comment by Ananya2001-an** [comment]
> Weirdly, I am losing notifications on GItHub. Thanks for answering, @Ananya2001-an.
> 
> Yes I agree, we need to break code the icon part into 3 sections (i.e break the prompt into two part as mentioned by Ananya)
> 
> ```js
> await prompt ([ ])
> await autocomplete.run()
> await prompt ([ ])
> 
> // the push the inputs
> 
> ```

Exactly just implement the same on all relevant functions and that should work.

**Comment by PratikDey** [comment]
Hi, while going by same logic as @Ananya2001-an mentioned, the `autocomplete.run()` is not running for the 2nd time and also it is taking a random value present.
image

**Comment by Pradumnasaraf** [comment]
> Hi, while going by same logic as @Ananya2001-an mentioned, the `autocomplete.run()` is not running for the 2nd time and also it is taking a random value present. image

Can you push the code? So that we can try and provide you the feedbcak

**Comment by PratikDey** [comment]
ye

> > Hi, while going by same logic as @Ananya2001-an mentioned, the autocomplete.run() is not running for the 2nd time and also it is taking a random value present.
> 
> Can you push the code? So that we can try and provide you with the feedback

Yeah done

**Comment by PratikDey** [comment]
Hi @Pradumnasaraf , I have pushed the code. You can review it.

**Comment by Pradumnasaraf** [comment]
Hey, @PratikDey, 

Thank you. Just allow me some time. A little bust with other works.

**Comment by Ananya2001-an** [comment]
Hey @PratikDey I tested the code and yes you are right that on the second time it's closing abruptly. I too am confused as to why it's happening. Let's wait for @Pradumnasaraf to give his review. Also thank you for being consistent on this till now. 😊

**Review by Pradumnasaraf — CHANGES_REQUESTED** [review]
Hey,

I think I got the root cause. You have created an object `const autocomplete = new AutoComplete({})` outside of the while loop and we are trying to update the `const` and an `object` in next iteration, so I think it's not possible.

So here is the solution, to put the object creation inside the loop

```js
async function addlinks(bool) {
  icons = await geticons();

  while (bool) {
    let answers = await prompt([
      {
        type: "input",
        name: "name",
        message: "What is the name of the link?",
      },
      {
        type: "input",
        name: "url",
        message: "What is the URL of the link?",
      }
    ]);

    const autocomplete = new AutoComplete({
      name: 'icon',
      message: 'Choose an icon (Search to see more options)',
      limit: 10,
      choices: icons
    })
    await autocomplete.run()

    const confirm = await prompt([
      {
        type: "confirm",
        name: "addLink",
        message: "Do you want to add another link?",
      }
    ])
    links.push({
      name: answers.name,
      url: answers.url,
      icon: autocomplete.state.input,
    });
    if (!confirm.addLink) {
      return links;
    }
  }
}
```

**Comment by Pradumnasaraf** [comment]
Also, fix this in `src/shared/questions/milestone.js`. It's causing error

```diff
- const { prompt } = require("enquirer");
+ const { promp, AutoComplete } = require("enquirer");
```

**Comment by Ananya2001-an** [comment]
> Hey,
> 
> I think I got the root cause. You have created an object `const autocomplete = new AutoComplete({})` outside of the while loop and we are trying to update the `const` and an `object` in next iteration, so I think it's not possible.
> 
> So here is the solution, to put the object creation inside the loop
> 
> ```js
> async function addlinks(bool) {
>   icons = await geticons();
> 
>   while (bool) {
>     let answers = await prompt([
>       {
>         type: "input",
>         name: "name",
>         message: "What is the name of the link?",
>       },
>       {
>         type: "input",
>         name: "url",
>         message: "What is the URL of the link?",
>       }
>     ]);
>   // solution
>     const autocomplete = new AutoComplete({
>       name: 'icon',
>       message: 'Choose an icon (Search to see more options)',
>       limit: 10,
>       choices: icons
>     })
>     await autocomplete.run()
> 
>     const confirm = await prompt([
>       {
>         type: "confirm",
>         name: "addLink",
>         message: "Do you want to add another link?",
>       }
>     ])
>     links.push({
>       name: answers.name,
>       url: answers.url,
>       icon: autocomplete.state.input,
>     });
>     if (!confirm.addLink) {
>       return links;
>     }
>   }
> }
> ```

That seems right. Since first we have selected an icon and its value is set inside the object it's not letting you change it because of `const` declaration.

**Comment by Pradumnasaraf** [comment]
> That seems right. Since first we have selected an icon and its value is set inside the object it's not letting you change it because of const declaration

Yes, I tried to use `let` but still doesn't work. I think we need a new object creation every time, so putting in that loop is the solution.

**Comment by Pradumnasaraf** [comment]
Hey, @PratikDey. You can implement this https://github.com/Pradumnasaraf/LinkFree-CLI/pull/53#pullrequestreview-1335916140. Ping us if you need help

**Comment by PratikDey** [comment]
Hi @Pradumnasaraf and @Ananya2001-an , I have made all the changes as requested. Still getting one issue in `socials.js` file.
When I am trying to update the socials already present in the json file, I am getting this error 
image

**Comment by Pradumnasaraf** [comment]
Hey,

Great. Can push the changes? So that we can have a look at the problem.

**Comment by Pradumnasaraf** [comment]
Oops. Gotcha. I missed the commit message. Thank you, we have a look.

File: src/shared/questions/links.js
Comment (Pradumnasaraf) [inline src/shared/questions/links.js:114]: Also, you don't need to add icons in this

**Review by Pradumnasaraf — CHANGES_REQUESTED** [review]

**Comment by Pradumnasaraf** [comment]
Also, @Ananya2001-an you can guide better in this because you have done the update part.

**Comment by Pradumnasaraf** [comment]
> > Also, @Ananya2001-an you can guide better in this because you have done the update part.
> 
> Sure I will look into it and tell u by tomorrow. Just caught up with some other work right now. BTW the changes you have pointed out are correct. They might the only reason for that issue earlier. Still I will recheck.

Sure, take your time. Yes, It's working after that change, but still need to check other things in that function.

**Comment by Ananya2001-an** [comment]
Hey @Pradumnasaraf I just tested the code right now and it works fine. I guess that was the only issue. @PratikDey You can follow up with what Pradumna has mentioned before and make the required changes.

**Comment by Pradumnasaraf** [comment]
Thanks for checking @Ananya2001-an .

Just a small question, we don't need to update anything in this part?
Screenshot 2023-03-15 at 12 50 27 PM

**Comment by Ananya2001-an** [comment]
> Thanks for checking @Ananya2001-an .
> 
> Just a small question, we don't need to update anything in this part?
> Screenshot 2023-03-15 at 12 50 27 PM
> 
> **Edit:** No, I don't think so because we are adding data to empty `let links = [];` 👍 
> 

Yes. I guess this is it.

**Comment by PratikDey** [comment]
Hi, even after doing the requested changes, the issue is still there.
Actually what is happening is that at the time of creating the json file, the socials are getting pushed into the `social` array and during update it is searching for `socials` array.

image

**Comment by PratikDey** [comment]
Committed all the changes, you folks can give this PR a review and let me know if any more changes are required.

**Comment by Pradumnasaraf** [comment]
Hey,

Yes, you pointed it out correctly. There is a bug in updating social. @Ananya2001-an can you have a look? 
I think after this fix we can merge this PR.

**Comment by Ananya2001-an** [comment]
> Hey,
> 
> Yes, you pointed it out correctly. There is a bug in updating social. @Ananya2001-an can you have a look? 
> I think after this fix we can merge this PR.

Sure I will look

**Comment by Ananya2001-an** [comment]
Hey @PratikDey just remove the changes made to updateJson file. It has to be `socials` only not `social`. Everything else is fine. @Pradumnasaraf you can also confirm.

**Comment by Pradumnasaraf** [comment]
> Hey @PratikDey just remove the changes made to updateJson file. It has to be `socials` only not `social`. Everything else is fine. @Pradumnasaraf you can also confirm.

Hey,

The issue is that it's not finding any existing socials.

Screenshot 2023-03-17 at 5 09 29 PM


**I think this is the issue**
Screenshot 2023-03-17 at 5 10 28 PM

**Comment by Ananya2001-an** [comment]
In updateJson there should be no changes made at all.

**Comment by Pradumnasaraf** [comment]
> In updateJson there should be no changes made at all.

Ah, gotcha. It's working now. My Gitpod caches the changes by Pratik's PR I have to delete are re-run. Thank you, Ananya

**Comment by Ananya2001-an** [comment]
No worries. Hey @PratikDey you can now make this last change so we can merge your PR.

**Comment by PratikDey** [comment]
I have reverted back to the old code in `updatejson`. Did not make any change to socialmap issue

**Review by Ananya2001-an — APPROVED** [review]
Looks good!👍🏼 @Pradumnasaraf give your review too.

**Review by Pradumnasaraf — APPROVED** [review]
Looks good to merge :tada: Thank you for your patience and contribution. Let's ship v2.2

**Pull request merged**
